### PR TITLE
[experiment] Remove PROGRAM_SIZE from type

### PR DESCRIPTION
**Pull request by mcalancea**
_naure:_
- Issue #566.
- Start a move towards circuit **parameters as values**, instead of as type.

### CODE REVIEW

File: ceno_zkvm/src/structs.rs [hunk 1]
```diff
@@ -152,10 +152,10 @@ impl<E: ExtensionField> ZKVMConstraintSystem<E> {
         config
     }
 
-    pub fn register_table_circuit<TC: TableCircuit<E>>(&mut self) -> TC::TableConfig {
+    pub fn register_table_circuit<TC: TableCircuit<E>>(&mut self, tc: TC) -> TC::TableConfig {
```
Comment (naure) [inline ceno_zkvm/src/structs.rs:155]: There could be a convenience method (original name) that uses Default (if implemented) and a new one with the parameter.

```rust
impl<TC> …
where TC::TableConfig: Default {
    pub fn register_table_circuit(…) {
        let tc = Default::default();
…
}

impl<TC> … {
    pub fn register_table_circuit_param(…, tc: TC) {
…
}
```

Comment (naure) [reply ceno_zkvm/src/structs.rs:155]: This way there is no need to change every call.

Comment (mcalancea) [reply ceno_zkvm/src/structs.rs]: :+1:  agreed, got a bit carried away there

File: ceno_zkvm/src/tables/program.rs [hunk 2]
```diff
@@ -104,11 +104,22 @@ pub struct ProgramTableConfig {
     mlt: WitIn,
 }
 
-pub struct ProgramTableCircuit<E, const PROGRAM_SIZE: usize>(PhantomData<E>);
+#[derive(Clone, Default)]
```
Comment (naure) [inline ceno_zkvm/src/tables/program.rs:107]: Default here is 0 size which is probably not what you want. Just leave it without default to force the use of the `…_param` function?

File: ceno_zkvm/src/tables/mod.rs [hunk 3]
```diff
@@ -18,18 +18,23 @@ pub use program::{InsnRecord, ProgramTableCircuit};
 mod ram;
 pub use ram::*;
 
-pub trait TableCircuit<E: ExtensionField> {
+pub trait TableCircuit<E: ExtensionField>
+where
+    Self: Default,
```
Comment (naure) [inline ceno_zkvm/src/tables/mod.rs:23]: This Default requirement could move as far out as possible towards the function that actually needs it.

**Review by naure — APPROVED** [review]

File: ceno_zkvm/examples/fibonacci_elf.rs [hunk 4]
```diff
@@ -79,15 +78,18 @@ fn main() {
...
     zkvm_cs.register_global_state::<GlobalState>();
 
     let mut zkvm_fixed_traces = ZKVMFixedTraces::default();
 
-    zkvm_fixed_traces.register_table_circuit::<ExampleProgramTableCircuit<E>>(
+    zkvm_fixed_traces.register_table_circuit_param::<ProgramTableCircuit<E>>(
         &zkvm_cs,
         &prog_config,
         vm.program(),
+        ptc_with_size,
```
Comment (naure) [inline ceno_zkvm/examples/fibonacci_elf.rs:92]: I applied the same change to this new file.

**Pull request closed without merging**